packages/flutter_tools/lib/src/compute_dev_dependencies.dart

```diff
@@ -0,0 +1,164 @@
+  } on FormatException catch (e) {
+    fail('$e');
+  }
+
+  List<T> asListOrFail<T>(Object? value, String name) {
+    if (value is! List<Object?>) {
+      fail('Expected field "$name" to be a list, got "$value"');
+    }
+    return <T>[
+      for (final Object? any in value)
+        if (any is T) any else fail('Expected element to be a $T, got "$any"')
+    ];
+  }
+
+  // Parse the JSON roughly in the following format:
+  //
+  // ```json
+  // {
+  //   "root": "my_app",
+  //   "packages": [
+  //     {
+  //       "name": "my_app",
+  //       "kind": "root",
+  //       "dependencies": [
+  //         "foo_plugin",
+  //         "bar_plugin"
+  //       ],
+  //       "directDependencies": [
+  //         "foo_plugin"
+  //       ],
+  //       "devDependencies": [
+  //         "bar_plugin"
+  //       ]
+  //     }
+  //   ]
+  // }
+  // ```
+  final List<Map<String, Object?>> packages = asListOrFail(
+    jsonResult['packages'],
+    'packages',
+  );
+
+  Map<String, Object?> packageWhere(
+    bool Function(Map<String, Object?>) test, {
+    required String reason,
+  }) {
+    return packages.firstWhere(test, orElse: () => fail(reason));
+  }
+
+  final Map<String, Object?> rootPackage = packageWhere(
+    (Map<String, Object?> package) => package['kind'] == 'root',
+    reason: 'A package with kind "root" was not found.',
+  );
+
+  // Start initially with every `devDependency` listed.
+  final Set<String> devDependencies = asListOrFail<String>(
+    rootPackage['devDependencies'],
+    'devDependencies',
+  ).toSet();
+
+  // Then traverse and exclude non-dev dependencies that list that dependency.
+  //
+  // This avoids the pathalogical problem of using, say, `path_provider` in a
+  // package's dev_dependencies:, but a (non-dev) dependency using it as a
+  // standard dependency - in that case we would not want to report it is used
+  // as a dev dependency.
+  final Set<String> visited = <String>{};
+  void visitPackage(String packageName) {
+    if (!visited.add(packageName)) {
+      return;
+    }
+
+    final Map<String, Object?> package = packageWhere(
+      (Map<String, Object?> package) => package['name'] == packageName,
+      reason: 'A package with name "$packageName" was not found',
+    );
+
+    // Do not traverse packages that themselves are dev dependencies.
+    if (package['kind'] == 'dev') {
+      return;
+    }
+
+    final List<String> directDependencies = asListOrFail(
+      package['directDependencies'],
+      'directDependencies',
+    );
+
+    // Remove any listed dependency from dev dependencies; it might have been
+    // a dev dependency for the app (root) package, but it is being used as a
+    // real dependency for a dependend on package, so we would not want to send
+    // a signal that the package can be ignored/removed.
+    devDependencies.removeAll(directDependencies);
+
+    // And continue visiting.
+    directDependencies.forEach(visitPackage);
```

Comment:
Is an infinite loop possible?

Reply:
No, clarified in both the comment and with a local variable above.